lj92: Fix issues related to SSSS=16

SimonSegerblomRex opened the pull request:
Hi!

I've been working on my own lossless JPEG encoder and found it useful to use the decoder in lj92 for debugging. (I don't know any details about the MLV format.) First I thought I had some bug in my encoder, but it turns out there are a few problems with the implementation of lj92 instead:
1. `SSSS=16` should only be used for `diff=-32678` as specified in the [spec](https://www.w3.org/Graphics/JPEG/itu-t81.pdf) in _H.1.2.2_.
2. `diff` values for `SSSS=16` should be encoded with zero bits (and decoded with zero bits just returning 32678 as the diff).

This pull-request fixes the first issue. It should be safe to merge with a couple of added benefits:
* Reduced size of the encoded buffer.
* Reduced risk of running into issues with decoders compliant with the spec. (From now on `SSSS=16` will not be used used for any bitdepths lower than 15.)

I don't know how you should handle the second issue. I can create a pull request fixing that as well, but I guess you need to keep both implementations and use some version tag in the metadata for controlling which decoder to use. (Adobe fixed the same issue for DNG 1.1, see _Compatibility Issue 2_ in _Appendix A_ of the [spec](https://www.adobe.com/content/dam/acom/en/products/photoshop/pdfs/dng_spec_1.4.0.0.pdf).)

bouncyball-git commented:
Thank you for pointing this out. Some long while ago when I was digging into this lib to adapt it to mlv_dump, mlvfs and mlvapp I also discovered that the author of lj92 lib (and mlrawplayer) has cut some corners. I fixed some small stuff to be sure it would work for out projects without errors, but it was a long time ago and I can't remember all of it now. In short the lj92 lib is quite messy and not the "best code" :) to be read and understood easily.

So we will definitely appreciate any good rework of it.

I'm not quite understand, why we need to maintain both implementations If you will fix the second issue?

bouncyball-git commented:
In addition, regarding to second fix, decoder compatibility depends on Canon HW encoder. 
We tried this and it works.
We have to check second fix against Canon HW encoded images.

SimonSegerblomRex commented:
> I'm not quite understand, why we need to maintain both implementations If you will fix the second issue?

Only if you want to support decoding images that have been encoded using the old implementation. Adobe suggests using the old implementation for DNG files with version <=1.1 (_Compatibility Issue 2 in Appendix A_ of the [spec](https://www.adobe.com/content/dam/acom/en/products/photoshop/pdfs/dng_spec_1.4.0.0.pdf).) I also discovered that the lossless JPEG decoder in ffmpeg suffers from the second issue (but since it only affects 15 and 16 bit images (where at least one diff value -32678) I guess noone noticed).

Should I upload the second fix as a new commit to this pull request, or will you merge this one first? (I'm not used to the GitHub workflow and not sure if I can create a new pull request depending on this one before it's merged (like in Gerrit).)

bouncyball-git commented:
diff = (int16_t)diff - works with Canon HW encoded images.

@masc4ii: can you also test this?

@SimonSegerblomRex: now it is show just one commit in your pull request. Can you put both commits (earlier and second  one) in this pull request?

SimonSegerblomRex commented:
> @SimonSegerblomRex: now it is show just one commit in your pull request. Can you put both commits (earlier and second one) in this pull request?

Sure! Done. (I just changed the commit message for the original commit and added the second commit.)

masc4ii commented:
> @masc4ii: can you also test this?

I can't find any errors. Works fine here... loaded some MLV and some transcoded CR2.

bouncyball-git commented:
Yup. Works fine here too. Could be safely merged?

masc4ii commented:
I think we could do.

SimonSegerblomRex commented:
I think it's ok to merge. You should see about 5 % decrease in file size when encoding.

masc4ii commented:
@SimonSegerblomRex : thank you very much! Merged.

ilia3101 commented:
Just saw this. Great to see!
